### ISSUE TITLE
Enhancement: [naming-convention] Allow number properties.

### ISSUE THREAD

**Issue by infacto** [typescript-eslint/typescript-eslint#5193]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md

### Description

![image](https://user-images.githubusercontent.com/60390085/173797547-e8b4f599-7243-4846-a9fe-57c387736c7c.png)

**.eslintrc.json** 

```json
{
  "@typescript-eslint/naming-convention": [
    "warn",
    {
      "selector": "objectLiteralProperty",
      "format": ["strictCamelCase"],
      "leadingUnderscore": "allow"
    },
  ]
}
```

Allow to have number keys for objects.

```ts
const example = {
  1: 'One',
  0x02: 'Two',
  0b11: 'Three'
}
```

**Comment by bradzacher** [comment]
The test for camelcase is based off three things:
1) the name must be a valid identifier.
2) the name must not contain underscores.
3) the first character must not be upper case.
4) Strict camelcase goes one step further and ensures that there are no two consecutive upper case characters.

A name that is a single number violates rules (1) and (3).
A name that is two or more numbers additionally violates rule (4).

Note that the check for uppercase is simply: `char === char.toUpperCase()`.

I don't agree that numbers should be special cased here.

[If you want to allow number keys, you can configure the rule to ignore them](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=MYewdgzgLgBApgDwIYFsAOAbOMC8MDeAUDDAPSkxgizAAWcwA1nACbEwCMAXDAOQDyYOLwA07AAwJxAJh68AKgHcQoiQCMO3PvNoAnOMLHtyMOg2ZsSSOapIBBG2Kt2HfVQF9CQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9GADa4bDkTRog6JAA08hVgXZIyREjL5V0yMRWloRDVt2QAZqq4d80ubvvJ8VYwGFuQwCDSDsFAF0wsABfTU8dT31DRGNTDHNLJRs4zygnaBc3DCYEeBzEh0p4fCVpBNyYRF5EAA8zAD0ZAAZaAE5w21y9QrIAC2kfWEQoiGiouYj5OeigA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by infacto** [comment]
Thanks, this is what I wanted. Ok, I wanted an option to allow number keys. But your solution also works. Before I created this issue, I tried putting the regex in addition to the existing property, thinking another definition would override it. But I see we can have multiple entries with the same selector. Is there some docs about it? We could add this (or similar) as example. Anyway... 👍 

```json
{
  "@typescript-eslint/naming-convention": [
    "warn",
    {
      "selector": "objectLiteralProperty",
      "format": ["strictCamelCase"],
      "leadingUnderscore": "allow"
    },
    {
      "selector": "property",
      "format": null,
      "filter": {
        "regex": "^[0-9]",
        "match": true
      }
    },
  ]
}
```
```js
const test = {
  0: 'Zero',
  0x01: 'One',
  0b10: 'Two',
  three: 'Three',
  42: 'Answer',
  foUR: 'Still not allowed.',
};
```

**Comment by bradzacher** [comment]
> Is there some docs about it?

https://typescript-eslint.io/rules/naming-convention/#how-does-the-rule-automatically-order-selectors

selectors are prioritised based on specificity and as soon as a matching selector is found the identifier is "complete".

The `filter` gives the second selector the highest priority, so it's evaluated first against all properties. Number properties always match it, so they must match `format: null` - which means "no format" (which is mentioned in [this section](https://typescript-eslint.io/rules/naming-convention/#format)).

Any non-number properties don't match the filter, so they don't match the selector, so the rule keeps looking for a selector that matches. The first selector has no filter, so all object literal properties will match it.
